internal/nginx/agent/config_store.go

```diff
@@ -0,0 +1,100 @@
+package agent
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/go-logr/logr"
+	"github.com/nginx/agent/sdk/v2/proto"
+
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/observer"
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/state/dataplane"
+)
+
+// NginxConfig is an intermediate object that contains nginx configuration in a form that agent expects.
+// We convert the dataplane configuration to NginxConfig in the config store, so we only need to do it once
+// per configuration change. The NginxConfig is then used by the agent to generate the nginx configuration payload.
+type NginxConfig struct {
+	ID          string
+	Config      *proto.ZippedFile
+	Aux         *proto.ZippedFile
+	Directories []*proto.Directory
+}
+
+// ConfigStore stores accepts the latest dataplane configuration and stores is as NginxConfig.
+// ConfigStore implements the observer.Subject interface,
+// so that it can notify the agent observers when the configuration changes.
+// ConfigStore is thread-safe.
+type ConfigStore struct {
```

Comment:
would it make sense for this type to be more generalized?  so that it doesn't deal with the NGK specific configuration. This way, it can potentially be reused in other projects, to configure a set of data plane replicas with the same configuration.

Reply:
How would that work? Would we just make this generic and have it store and return the `any` type?

I'm worried about extracting the building piece of this. I only want to build the nginx agent config once because it involves costly operations like checksums.

Reply:
> I guess I could perform the dataplane.Configuration -> agent.NginxConfiguration before storing the config and notifying the observers.

yep